v1/plugin/stream_collector_proxy.go

```diff
@@ -1,119 +1,177 @@
 package plugin
 
 import (
+	"errors"
 	"fmt"
+	"time"
 
 	"golang.org/x/net/context"
 
 	"github.com/intelsdi-x/snap-plugin-lib-go/v1/plugin/rpc"
 )
 
+const defaultMaxCollectDuration = 10 * time.Second
+
 type StreamProxy struct {
 	pluginProxy
-
 	plugin StreamCollector
+
+	// maxMetricsBuffer is the maximum number of metrics the plugin should buffer
+	// before sending metrics, defaults to zero what means send metrics immediately
+	maxMetricsBuffer int64
+
+	// maxCollectionDuration sets the maximum duration between collections before metrics should be sent
+	// (defaults to 10s what means that after 10 seconds no new metrics are received,
+	// the plugin should send whatever data it has in buffer instead of waiting longer)
+	maxCollectDuration time.Duration
 }
 
-func (c *StreamProxy) StreamMetrics(stream rpc.StreamCollector_StreamMetricsServer) error {
+func (p *StreamProxy) GetMetricTypes(ctx context.Context, arg *rpc.GetMetricTypesArg) (*rpc.MetricsReply, error) {
+	cfg := fromProtoConfig(arg.Config)
+
+	r, err := p.plugin.GetMetricTypes(cfg)
+	if err != nil {
+		return nil, err
+	}
+	metrics := []*rpc.Metric{}
+	for _, mt := range r {
+		// We can ignore this error since we are not returning data from
+		// GetMetricTypes.
+		metric, _ := toProtoMetric(mt)
+		metrics = append(metrics, metric)
+	}
+	reply := &rpc.MetricsReply{
+		Metrics: metrics,
+	}
+	return reply, nil
+}
+
+func (p *StreamProxy) StreamMetrics(stream rpc.StreamCollector_StreamMetricsServer) error {
+	if stream == nil {
+		return errors.New("Stream metrics server is nil")
+	}
+
 	// Error channel where we will forward plugin errors to snap where it
 	// can report/handle them.
 	errChan := make(chan string)
 	// Metrics into the plugin from snap.
 	inChan := make(chan []Metric)
 	// Metrics out of the plugin into snap.
 	outChan := make(chan []Metric)
-	err := c.plugin.StreamMetrics(inChan, outChan, errChan)
+
+	err := p.plugin.StreamMetrics(inChan, outChan, errChan)
 	if err != nil {
 		return err
 	}
-	go metricSend(c.plugin, outChan, stream)
-	go errorSend(c.plugin, errChan, stream)
-	streamRecv(c.plugin, inChan, stream)
+
+	go p.metricSend(outChan, stream)
+	go p.errorSend(errChan, stream)
+	p.streamRecv(inChan, stream)
+
 	return nil
 }
 
-func errorSend(
-	plugin StreamCollector,
-	ch chan string,
-	s rpc.StreamCollector_StreamMetricsServer) {
-	for r := range ch {
+func (p *StreamProxy) SetConfig(context.Context, *rpc.ConfigMap) (*rpc.ErrReply, error) {
+	return nil, nil
+}
```

Comment:
To reviewers:
Please notice that this `SetConfig` is not added in this PR - it's only moved from [here](https://github.com/intelsdi-x/snap-plugin-lib-go/pull/77/files#diff-aa6fc80eb3b8c2400af503a0443c72b7L97)